libkbfs/folder_branch_ops.go

```diff
@@ -3593,6 +3624,45 @@ func (fbo *folderBranchOps) getAndApplyMDUpdates(ctx context.Context,
 	return nil
 }
 
+func (fbo *folderBranchOps) getAndApplyNewestUnmergedHead(ctx context.Context,
+	lState *lockState) error {
+	fbo.log.CDebugf(ctx, "Fetching the newest unmerged head")
+	bid := func() BranchID {
+		fbo.mdWriterLock.Lock(lState)
+		defer fbo.mdWriterLock.Unlock(lState)
+		return fbo.bid
+	}()
+
+	// We can only ever be at most one revision behind, so fetch the
+	// latest unmerged revision and apply it as a successor.
+	md, err := fbo.config.MDOps().GetUnmergedForTLF(ctx, fbo.id(), bid)
+	if err != nil {
+		return err
+	}
+
+	if md == (ImmutableRootMetadata{}) {
+		// There is no unmerged revision, oops!
+		return errors.New("Couldn't find an unmerged head")
+	}
+
+	fbo.mdWriterLock.Lock(lState)
+	defer fbo.mdWriterLock.Unlock(lState)
+	if fbo.isMasterBranchLocked(lState) {
```

Comment:
Isn't it possible (although unlikely) that we switch back to the master branch, but then get unmerged again? Perhaps just check `bid` again and return if they differ?